### PR TITLE
AST rewrite support for String Template - JEP459

### PR DESCRIPTION
**Pull request by subyssurendran666**
Fixes https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1800

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by subyssurendran666** [comment]
WIP

**Comment by subyssurendran666** [comment]
@mpalat, could you please review this PR.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 1]
```diff
@@ -665,8 +666,13 @@ private int rewriteList(
 						if (separatorState == EXISTING) {
 							updateIndent(prevMark, currPos, i, editGroup);
 						}
-
-						doTextInsert(currPos, node, getNodeIndent(i), true, editGroup); // insert node
+							doTextInsert(
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:669]: this function is a much used function and we should avoid node-specific checks here. need to hoist this part of checking for StringTeemplate at the caller side isolated in that node - or find another way to do this rewrite.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 2]
```diff
@@ -913,6 +919,42 @@ private int rewriteJavadoc(ASTNode node, StructuralPropertyDescriptor property)
...
+				case RewriteEvent.REPLACED: {
+					boolean originalNode= ((Boolean) event.getOriginalValue()).booleanValue();
+					boolean newValue= ((Boolean) event.getNewValue()).booleanValue();
+
+					int pos1 = rewriteRequiredNode(node, StringTemplateExpression.TEMPLATE_PROCESSOR);
+					int pos2 = rewriteRequiredNode(node, StringTemplateExpression.STRING_TEMPLATE_COMPONENTS);
+
+					if(newValue && !originalNode) {// SINGLE LINE to MULTI LINE
+						if(pos2 == 0 ) { // without component
+							doTextReplace(pos1 +1, 1, "\"\"\"\n", getEditGroup(event)); //$NON-NLS-1$
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:936]: nitpicking - put a single String for these quotations and use that variable here for "\"\"\n" etc

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 3]
```diff
@@ -913,6 +919,42 @@ private int rewriteJavadoc(ASTNode node, StructuralPropertyDescriptor property)
...
+							doTextReplace(pos -1, 1, "\n\"\"\"", getEditGroup(event)); //$NON-NLS-1$
+						} else { // with Component
+							doTextReplace(pos1 +1, 1, "\"\"\"\n", getEditGroup(event)); //$NON-NLS-1$
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:939]: this code can be simplified to be one with appropriate pos assignments

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 4]
```diff
@@ -913,6 +919,42 @@ private int rewriteJavadoc(ASTNode node, StructuralPropertyDescriptor property)
...
+							doTextReplace(pos2, 1, "\n\"\"\"", getEditGroup(event)); //$NON-NLS-1$
+						}
+
+					} else if(!newValue && originalNode) {// MULTI LINE to SINGLE LINE
+						if(pos2 == 0) { // without Component
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:944]: same comments as above

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 5]
```diff
@@ -1424,7 +1466,7 @@ private int getPosAfterLeftBrace(int pos) {
 	 */
 	private int getPosAfterRightParenthesis(int pos) {
 		try {
-			return getPosAfterToken(pos, TerminalTokens.TokenNameRPAREN);
+				return getPosAfterToken(pos, TerminalTokens.TokenNameRPAREN);
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:1469]: please cleanup here

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 6]
```diff
@@ -4677,11 +4718,20 @@ public boolean visit(StringTemplateExpression node) {
...
+
 		if (node.getAST().isPreviewEnabled()) {
-			int pos = rewriteRequiredNode(node, StringTemplateExpression.TEMPLATE_PROCESSOR);
-			pos = rewriteRequiredNode(node, StringTemplateExpression.FIRST_STRING_FRAGMENT);
-			rewriteNodeList(node, StringTemplateExpression.STRING_TEMPLATE_COMPONENTS, pos, "", "", ""); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+			rewriteNodeList(
+					node,
+					StringTemplateExpression.STRING_TEMPLATE_COMPONENTS,
+					//StringTemplateExpression.MULTI_LINE,
+					node.components().isEmpty() ? pos -1 : pos,//Sometimes, the pos  returns false values even when the stringTemplateExpression has no stringTemplateComponent
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:4728]: what is the meaning of this comment?

**Review by mpalat — COMMENTED** [review]
please resubmit after incorporating the review comments

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 1 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:677]: Done.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 5 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 2 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 4 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:958]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 6 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 3 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java]: Done

**Comment by mpalat** [comment]
@subyssurendran666 Please wait for https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2117 to be merged and  then rebase. The test failures due to merge of some new features are being addressed by that PR.

**Review by mpalat — APPROVED** [review]
+1

**Comment by mpalat** [comment]
Thanks @subyssurendran666

**Comment by mpalat** [comment]
done via https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2026

**Pull request merged**